relayer/chains/wasm/provider.go

```diff
@@ -706,78 +707,78 @@ func (p *Provider) runBlockQuery(ctx context.Context, blockInfoChan chan *relayT
-		p.logger.Error("http client start failed", zap.Error(err))
-		return err
-	}
-	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	query := strings.Join([]string{
 		"tm.event = 'Tx'",
 		fmt.Sprintf("tx.height >= %d ", opts.Height),
 		fmt.Sprintf("%s._contract_address = '%s'", opts.Method, opts.Address),
 	}, " AND ")
 
-	resultEventChan, err := httpClient.Subscribe(newCtx, opts.Address, query)
+	resultEventChan, err := p.client.Subscribe(ctx, "tx-result-waiter", query)
 	if err != nil {
 		p.logger.Error("event subscription failed", zap.Error(err))
 		return p.SubscribeMessageEvents(ctx, blockInfoChan, opts)
 	}
-	defer httpClient.Unsubscribe(ctx, opts.Address, query)
-	p.logger.Info("event subscription started")
+	defer p.client.Unsubscribe(ctx, opts.Address, query)
+	p.logger.Info("event subscription started", zap.String("contract_address", opts.Address), zap.String("method", opts.Method))
 
 	for {
 		select {
 		case <-ctx.Done():
 			p.logger.Info("event subscription stopped")
 			return ctx.Err()
 		case e := <-resultEventChan:
-			p.logger.Info("event received")
 			eventDataJSON, err := jsoniter.Marshal(e.Data)
 			if err != nil {
 				p.logger.Error("failed to marshal event data", zap.Error(err))
 				continue
 			}
-			eventsList := &struct {
-				Height uint64  `json:"height"`
-				Events []Event `json:"logs"`
-			}{}
-			if err := jsoniter.Unmarshal(eventDataJSON, eventsList); err != nil {
+			var res types.TxResultWaitResponse
+			if err := jsoniter.Unmarshal(eventDataJSON, &res); err != nil {
 				p.logger.Error("failed to unmarshal event data", zap.Error(err))
 				continue
 			}
-			messages, err := p.ParseMessageFromEvents(eventsList.Events)
-			if err != nil {
-				p.logger.Error("failed to parse message from events", zap.Error(err))
+			eventsList := []struct {
+				Events []Event `json:"events"`
+			}{}
+			if err := jsoniter.Unmarshal([]byte(res.Result.Log), &eventsList); err != nil {
+				p.logger.Error("failed to unmarshal event list", zap.Error(err))
 				continue
 			}
+			var messages []*relayTypes.Message
+			for _, event := range eventsList {
+				msgs, err := p.ParseMessageFromEvents(event.Events)
+				if err != nil {
+					p.logger.Error("failed to parse message from events", zap.Error(err))
+					continue
+				}
+				messages = append(messages, msgs...)
+			}
+
 			blockInfo := &relayTypes.BlockInfo{
-				Height:   eventsList.Height,
+				Height:   uint64(res.Height),
 				Messages: messages,
 			}
 			blockInfoChan <- blockInfo
+			opts.Height = blockInfo.Height
 			for _, msg := range blockInfo.Messages {
 				p.logger.Info("Detected eventlog",
-					zap.Uint64("height", eventsList.Height),
+					zap.Int64("height", res.Height),
 					zap.String("target_network", msg.Dst),
 					zap.Uint64("sn", msg.Sn),
 					zap.String("event_type", msg.EventType),
 				)
 			}
-			opts.Height = eventsList.Height
 		default:
-			if httpClient.IsRunning() {
-				continue
+			if !p.client.IsConnected() {
```

Comment:
When a WebSocket connection is closed or disconnected, there's typically a delay in reconnection. During this interval, real-time events may be missed, possibly including entire blocks of data. I strongly believe this situation should be handled in a reliable way.